netcdftime/_netcdftime.pyx

```diff
@@ -1151,8 +1156,470 @@ def time2index(times, nctime, calendar=None, select='exact'):
     return _toscalar(index)
 
 
-def _toscalar(a):
+cdef _toscalar(a):
     if a.shape in [(), (1,)]:
         return a.item()
     else:
         return a
+
+cdef to_tuple(dt):
+    """Convert a datetime instance into a tuple of integers. Elements go
+    in the order of decreasing significance, making it easy to compare
+    datetime instances. Parts of the state that don't affect ordering
+    are omitted. Compare to datetime.timetuple()."""
+    return (dt.year, dt.month, dt.day, dt.hour, dt.minute,
+            dt.second, dt.microsecond)
+
+# a cache of converters (utime instances) for different calendars
+cdef dict _converters
+_converters = {}
+for calendar in _calendars:
+    _converters[calendar] = utime("seconds since 1-1-1", calendar)
+
+cdef class datetime(object):
+
+    """
+Phony datetime object which mimics the python datetime object,
+but allows for dates that don't exist in the proleptic gregorian calendar.
+Doesn't do timedelta operations, doesn't overload + and -.
+
+Has strftime, timetuple and __repr__ methods.  The format
+of the string produced by __repr__ is controlled by self.format
+(default %Y-%m-%d %H:%M:%S). Does support comparisons with other
+phony datetime and with datetime.datetime objects.
+
+Instance variables are year,month,day,hour,minute,second,dayofwk,dayofyr
+and format.
+    """
+    cdef readonly int year, month, day, hour, minute, dayofwk, dayofyr
+    cdef readonly int second, microsecond
+    cdef readonly str calendar
+
+    # Python's datetime.datetime uses the proleptic Gregorian
+    # calendar. This boolean is used to decide whether a
+    # netcdftime.datetime instance is comparambe to datetime.datetime.
+    cdef readonly bint calendar_is_gregorian
+
+    # The pointer to a function used by __add__() and __sub() for
+    # datetime + timedelta computations. This allows us to check the
+    # calendar once (in __init__()) instead of doing it every time
+    # __add__() and __sub__ are called.
+    cdef datetime (*_add_timedelta)(datetime, object)
```

Comment:
OK. Please see the new version, now using inheritance instead of a function pointer.